File: code/python/06.2.1_Alertor/Alertor.py
#!/usr/bin/env python3
########################################################################################
# Filename    : Alertor.py
# Description : Make Alertor with buzzer and button
# Author      : www.freenove.com
# modification: 2019/12/27
########################################################################################

########################################################################################
##
## import `TonalBuzzer`, `Button`, and `Tone` functionality from `gpiozero` library
##
from gpiozero import TonalBuzzer,Button
from gpiozero.tones import Tone

########################################################################################
##
## import the `time` and `math` libs
##
import time
import math

########################################################################################
##
## instantiating the TonalBuzzer class on pin 17
##
buzzer = TonalBuzzer(17)

########################################################################################
##
## instantiating the Button class on pin 18
##
button = Button(18)

def loop():
    while True:
        
        ################################################################################
        ##
        ## if button is pressed
        ##
        if button.is_pressed:
        
            ############################################################################
            ##
            ## start playing a repeating tone
            ##
            alertor()
            
            ############################################################################
            ##
            ## display operational information
            ##
            print ('alertor turned on >>> ')
            
        else:
        
            ############################################################################
            ##
            ## stop playing a repeating tone
            ##
            stopAlertor()
            
            ############################################################################
            ##
            ## display operational information
            ##
            print ('alertor turned off <<<')
            
########################################################################################
##
## start playing a repeating tone
##            
def alertor():
    
    ####################################################################################
    ##
    ## play a tone using the buzzer
    ##
    buzzer.play(Tone(220.0))
    
    ####################################################################################
    ##
    ## strategic delay for user perception (.1 seconds)
    ##
    time.sleep(0.1)
        
def stopAlertor():

    ####################################################################################
    ##
    ## stop playing the buzzer tone
    ##
    buzzer.stop()
            
def destroy():
    
    ####################################################################################
    ##
    ## closes the connection to the buzzer pin
    ##
    buzzer.close()                  

########################################################################################
##
## start of program
##
if __name__ == '__main__':

    ####################################################################################
    ##
    ## display operational information
    ##
    print ('Program is starting...')
    
    ####################################################################################
    ##
    ## normal operations: run the loop
    ##
    try:
        loop()
        
    ####################################################################################
    ##
    ## catch any keyboard interrupt characters to signal program exit
    ##
    except KeyboardInterrupt:  # Press ctrl-c to end the program.
        destroy()
        print("Ending program")

File: code/C++/06.2.1_alertor/alertor.c
//
// Filename:    alertor.c
// Description: Make Alertor with BUZZER and BUTTON
// Components:  BUTTON, BUZZER
//
////////////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////////////
//
// Pre-processor directives: include header files for wiringPi and math libraries,
// define symbols for easier reference to our utilized wiringPi GPIOs
//
#include <math.h>
#include <softTone.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <wiringPi.h>

#define  BUTTONpin     1      // define the BUTTONpin
#define  BUZZERpin     0      // define the BUZZERpin

////////////////////////////////////////////////////////////////////////////////////////
//
// Function prototypes for buzzer operation
//
void     alertor     (int32_t);
void     stopAlertor (int32_t);

int32_t  main (void)
{
    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Bring wiringPi functionality online
    //
    if (wiringPiSetup ()            == -1)
    {
        fprintf (stderr, "[ERROR] Could not initialize wiringPi library!\n");
        exit (1);
    }

    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Set pin modes for involved wiringPi GPIO pins, and configure button and buzzer
    //
    pinMode (BUTTONpin, INPUT);
    pinMode (BUZZERpin, OUTPUT);
    softToneCreate (BUZZERpin);           // set BUZZERpin for tone generation
    pullUpDnControl (BUTTONpin, PUD_UP);  // pull up to HIGH level

    fprintf (stdout, "Program is starting (CTRL-c to interrupt) ...\n");
    fprintf (stdout, "[ALERTOR] turned OFF");

    while (1)
    {
        ////////////////////////////////////////////////////////////////////////////////
        //
        // Button is pressed, generate alert using buzzer
        //
        if (digitalRead (BUTTONpin) == LOW)
        {
            alertor (BUZZERpin);
            fprintf (stdout, "\b\b\bON ");
        }

        ////////////////////////////////////////////////////////////////////////////////
        //
        // Button is released, quiet the buzzer
        //
        else
        {
            stopAlertor (BUZZERpin);
            fprintf (stdout, "\b\b\bOFF");
        }
    }

    return (0);
}

////////////////////////////////////////////////////////////////////////////////////////
//
// Generate an alert from the buzzer on indicated pin
//
void alertor (int32_t  pin)
{
    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Declare and initialize variables
    //
    int32_t  x        = 0;
    double   sinVal   = 0.0;
    double   toneVal  = 0.0;

    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Calculate and generate tone, utilizing the sine function
    //
    for (x = 0; x < 360; x++)
    {
        sinVal        = sin (x * (M_PI / 180)); // Calculate the sine value
        toneVal       = 2000 + sinVal * 500;    // Add frequency, weighted sine value
        softToneWrite (pin, toneVal);          // output corresponding PWM
        delay (1);
    }
}

////////////////////////////////////////////////////////////////////////////////////////
//
// Quiet the buzzer on indicated wiringPi GPIO pin
//
void stopAlertor (int32_t  pin)
{
    softToneWrite (pin, 0);
}
